File: solutions/33/recursive_duplicates.js
'use strict'; const log = console.log; (async ()=>{

/* f :: [Double] -> Double -> Int */
const g = (A, x) => f(A, 0, A.length - 1, x);

const f = function p(A, l, h, x) {
  if(l > h) return -1;

  const m = l + Math.trunc((h - l) / 2);
  if(A[m] === x) return m;

  if(A[l] < A[m])
    if(A[l] <= x && x < A[m])
      return p(A, l, m - 1, x);
    else
      return p(A, m + 1, h, x);
  else if(A[m] < A[h])
    if(A[m] < x && x <= A[h])
      return p(A, m + 1, h, x);
    else
      return p(A, l, m - 1, x);
  else {
      const y = p(A, l, m - 1, x);
      if(y === -1)
        return p(A, m + 1, h, x);
      else
        return y;
  }

  return -1;
};

const eq = require('assert').strictEqual;
eq(g([], 5), -1);
eq(g([-1, 0, 1, 2, 5, 7, -5, -3], -5), 6);
eq(g([-1, 0, 1, 2, 5, 7, -5, -3], -2), -1);
eq(g([2, 2, 2, 3, 4, 2], -5), -1);
eq(g([2, 2, 2, 3, 4, 2], 2), 2);
eq(g([2, 2, 2, 3, 4, 2], 3), 3);
eq(g([2, 2, 2, 3, 4, 2], 4), 4);

eq(g([3, 4, 2, 2, 2, 2], -5), -1);
eq(g([3, 4, 2, 2, 2, 2], 2), 2);
eq(g([3, 4, 2, 2, 2, 2], 4), 1);
eq(g([3, 4, 2, 2, 2, 2], 3), 0);
eq(g([3, 4, 1, 2, 2, 2], 3), 0);
eq(g([2, 4, 2, 2, 2, 2], 4), 1);

})();

File: solutions/33/recursive_duplicates.py
def f(A, k):
    if not A:
        return -1

    return g(A, k, 0, len(A) - 1)

def g(A, k, l, r):
    if l > r:
        return -1

    m = l + (r - l) // 2

    if A[m] == k:
        return m

    if A[l] < A[m]:
        if A[l] <= k < A[m]:
            return g(A, k, l, m - 1)
        else:
            return g(A, k, m + 1, r)
    elif A[m] < A[r]:
        if A[m] < k <= A[r]:
            return g(A, k, m + 1, r)
        else:
            return g(A, k, l, m - 1)
    else:
        x = g(A, k, l, m - 1)
        if x < 0:
            x = g(A, k, m + 1, r)
        return x

assert f([], 5) == -1
assert f([-1, 0, 1, 2, 5, 7, -5, -3], -5) == 6
assert f([-1, 0, 1, 2, 5, 7, -5, -3], -2) == -1
assert f([2 , 4, 2, 2, 2, 2], 4) == 1
assert f([5], 5) == 0
assert f([-2, -1, 0, 1, 2, 3, 4, 5], 4) == 6
assert f([2, 2, 2, 3, 4, 2], -5) == -1
assert f([2, 2, 2, 3, 4, 2], 2) == 2
assert f([2, 2, 2, 3, 4, 2], 3) == 3
assert f([2, 2, 2, 3, 4, 2], 4) == 4
assert f([3, 4, 2, 2, 2, 2], -5) == -1
assert f([3, 4, 2, 2, 2, 2], 2) == 2
assert f([3, 4, 2, 2, 2, 2], 4) == 1
assert f([3, 4, 2, 2, 2, 2], 3) == 0
assert f([3, 4, 1, 2, 2, 2], 3) == 0
assert f([2, 4, 2, 2, 2, 2], 4) == 1
